psydac/fem/tensor.py

```diff
@@ -183,45 +191,62 @@ def eval_field( self, field, *eta, weights=None):
         if weights:
             assert weights.space == field.coeffs.space
 
-        bases = []
         index = []
+        if weights:
+            w_index = []
+
+        if not hasattr(self,'_tmp_coeffs'):
+            self._tmp_coeffs = np.zeros([s.degree +1 for s in self.spaces])
+        if not hasattr(self,'_tmp_bf'):
+            self._tmp_bf = [np.zeros(s.degree + 1, dtype=float) for s in self.spaces]
+
+        coeffs = self._tmp_coeffs
 
         # Necessary if vector coeffs is distributed across processes
         if not field.coeffs.ghost_regions_in_sync:
             field.coeffs.update_ghost_regions()
 
-        for (x, xlim, space) in zip( eta, self.eta_lims, self.spaces ):
+        for i, (x, xlim, space) in enumerate(zip( eta, self.eta_lims, self.spaces )):
 
             knots  = space.knots
             degree = space.degree
-            span   =  find_span( knots, degree, x )
-
+            # be sure it's the right type for kernels
+            fx = float(x)
+            span   =  find_span_p( knots, degree, fx )
+            vs = space.vector_space
             #-------------------------------------------------#
             # Fix span for boundaries between subdomains      #
             #-------------------------------------------------#
             # TODO: Use local knot sequence instead of global #
             #       one to get correct span in all situations #
             #-------------------------------------------------#
-            if x == xlim[1] and x != knots[-1-degree]:
+            if fx == xlim[1] and fx != knots[-1-degree]:
                 span -= 1
             #-------------------------------------------------#
-            basis = basis_funs( knots, degree, x, span)
+            
+            basis = self._tmp_bf[i]
+            basis_funs_p( knots, degree, fx, span, out = basis)
 
             # If needed, rescale B-splines to get M-splines
             if space.basis == 'M':
                 basis *= space.scaling_array[span-degree : span+1]
 
             # Determine local span
-            wrap_x   = space.periodic and x > xlim[1]
+            wrap_x   = space.periodic and fx > xlim[1]
             loc_span = span - space.nbasis if wrap_x else span
-
-            bases.append( basis )
-            index.append( slice( loc_span-degree, loc_span+1 ) )
+            start = vs.starts[0]
+            pad = vs.pads[0]
+            shift = vs.shifts[0]
+            index.append( slice( loc_span-degree-start+shift*pad, loc_span+1-start+shift*pad ) )
+            if weights:
+                w_index.append(slice(loc_span-degree, loc_span+1))
         # Get contiguous copy of the spline coefficients required for evaluation
+        bases = self._tmp_bf
         index  = tuple( index )
-        coeffs = field.coeffs[index].copy()
+        coeffs[:] = field.coeffs._data[index]        
```

Comment:
Accessing `coeffs._data` instead of just `coeffs` does not appear to work in all cases (see failing tests).

Reply:
This appears to work now.